### PR TITLE
Issue #85: add support for new constraint operators

### PR DESCRIPTION
**Pull request by daveleek**
# Description

Adds support for additional constraint operators: 

Semantic Versioning
SEMVER_EQ
SEMVER_GT
SEMVER_LT

Numbers
NUM_EQ
NUM_GT
NUM_GTE
NUM_LT
NUM_LTE

Date
DATE_AFTER
DATE_BEFORE

Strings
STR_ENDS_WITH
STR_STARTS_WITH
STR_CONTAINS

Also adds inverted feature to constraints

Fixes # (issue)
#85 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- Added a series of unittests to verify functionality
- Ran the spec tester

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/Unleash/Strategies/ConstraintUtils.cs [hunk 1]
```diff
@@ -21,8 +39,19 @@ public static bool Validate(List<Constraint> constraints, UnleashContext context
...
+                return operators[constraint.Operator].Evaluate(constraint, context);
+            else
+            {
+                var isIn = contextValue != null && constraint.Values.Contains(contextValue.Trim());
+
+                if (constraint.Operator == Operator.IN)
+                    return isIn;
+                if (constraint.Operator == Operator.NOT_IN)
+                    return !isIn;
+            }
```
Comment (daveleek) [inline src/Unleash/Strategies/ConstraintUtils.cs:52]: I'll move this into the StringConstraintOperator

File: src/Unleash/Strategies/Constraints/SemverConstraintOperator.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            var contextValue = context.GetByName(constraint.ContextName);
+            SemanticVersion contextSemver;
+            if (!SemanticVersion.TryParse(contextValue, out contextSemver))
+            {
+                Logger.Info("Couldn't parse version {0} from context");
+                return false;
+            }
+
+            if (constraint.Values != null && constraint.Values.Any())
```
Comment (daveleek) [inline src/Unleash/Strategies/Constraints/SemverConstraintOperator.cs:25]: After checking the spec and checking with @chriswk it's decided that i should switch to only .Value here

File: tests/Unleash.Tests/Integration/UnleashContextDefinition.cs [hunk 3]
```diff
@@ -9,6 +10,7 @@ public class UnleashContextDefinition
         public string RemoteAddress { get; set; }
         public string Environment { get; set; }
         public string AppName { get; set; }
+        public DateTimeOffset CurrentTime { get; set; }
```
Comment (daveleek) [inline tests/Unleash.Tests/Integration/UnleashContextDefinition.cs:13]: Forgot to make this nullable

File: src/Unleash/Strategies/Constraints/SemverConstraintOperator.cs (hunk 2 above)
Comment (chriswk) [reply src/Unleash/Strategies/Constraints/SemverConstraintOperator.cs:25]: :+1:

**Review by chriswk — APPROVED** [review]
Looks really good. Thanks for the great work.

Feel free to squash the commits so we get a clean commit adding this

**Pull request merged**